internal/pkg/docker/orchestrator/orchestrator.go

```diff
@@ -452,6 +486,78 @@ func (o *Orchestrator) waitForContainerToStart(ctx context.Context, id string) e
 	}
 }
 
+func (o *Orchestrator) waitForContainerDependencies(ctx context.Context, name string, definitions map[string]ContainerDefinition) error {
+	var deps []string
+	for depName, state := range definitions[name].DependsOn {
+		deps = append(deps, fmt.Sprintf("%s->%s", depName, state))
+	}
+	logMsg := strings.Join(deps, ", ")
+	fmt.Printf("Waiting for container %q dependencies: [%s]\n", name, color.Emphasize(logMsg))
+	eg, ctx := errgroup.WithContext(ctx)
+	for name, state := range definitions[name].DependsOn {
+		name, state := name, state
+		eg.Go(func() error {
+			ctrId := o.containerID(name)
+			ticker := time.NewTicker(700 * time.Millisecond)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+				switch state {
+				case ctrStateStart:
+					var errContainerExited *dockerengine.ErrContainerExited
+					err := o.waitForContainerToStart(ctx, ctrId)
+					if err != nil && !errors.As(err, &errContainerExited) {
+						return fmt.Errorf("wait for container %q to start: %w", ctrId, err)
+					}
+					log.Successf("Successfully started container %s\n", ctrId)
```

Comment:
I was at first wondering if this was necessary, because I had assumed that we would start processing this vertex (container) if all of its previous vertices have been processed - if they have been processed, that means they've started.

Now I realized that we will start processing this vertex right after all of its ancestors have entered "visited" or "visiting", which means that it's not guaranteed that all of the ancestor containers have been started.

This led me to the question here: https://github.com/aws/copilot-cli/pull/5509#discussion_r1431789758.

Edit:
I see that we've updated `graph` such that we process a vertex only if all of the predecessors have been "visited". Does that imply that all the predecessor containers must have all been started successfully (though completion and health state are still unknown)?

Reply:
Absolutely right 👍 As we modified graph pkg. This block of code is now unnecessary. removed it!